lib/internal/loader/DefaultResolve.js

```diff
@@ -0,0 +1,90 @@
+'use strict';
+
+const { URL } = require('url');
+const CJSmodule = require('module');
+const internalURLModule = require('internal/url');
+const internalFS = require('internal/fs');
+const NativeModule = require('native_module');
+const { extname } = require('path');
+const { realpathSync } = require('fs');
+const preserveSymlinks = !!process.binding('config').preserveSymlinks;
+const errors = require('internal/errors');
+const { resolve: moduleWrapResolve } = internalBinding('module_wrap');
+
+const realpathCache = new Map();
+
+function search(target, base) {
+  if (base === undefined) {
+    // We cannot search without a base.
+    throw new errors.Error('ERR_MISSING_MODULE', target);
+  }
+  try {
+    return moduleWrapResolve(target, base);
+  } catch (e) {
+    e.stack; // cause V8 to generate stack before rethrow
+    let error = e;
+    try {
+      const questionedBase = new URL(base);
+      const tmpMod = new CJSmodule(questionedBase.pathname, null);
+      tmpMod.paths = CJSmodule._nodeModulePaths(
+        new URL('./', questionedBase).pathname);
+      const found = CJSmodule._resolveFilename(target, tmpMod);
+      error = new errors.Error('ERR_MODULE_RESOLUTION_LEGACY', target,
+                               base, found);
+    } catch (problemChecking) {
+      // ignore
+    }
+    throw error;
+  }
+}
+
+function resolve(specifier, parentURL) {
+  if (NativeModule.nonInternalExists(specifier)) {
+    return {
+      url: specifier,
+      format: 'builtin'
+    };
+  }
+
+  let url;
+  try {
+    url = search(specifier, parentURL);
+  } catch (e) {
+    if (e.message && e.message.startsWith('Cannot find module'))
```

Comment:
should this be made robust against modification of `String.prototype.startsWith`?

Also, please add a test case for an exception that has a non-string `message` property.